### PR TITLE
Make URL form setting consistent with URL action calling

### PR DESCRIPTION
**Pull request by zaucker**
Allow form entries from URL with
```
        {
            key    => 'nevis_test',
            label  => trm('Test'),
            widget => 'text',
            set    => {
                required => true,
                maxWidth => 300,
            },
            urlFormKey => 'test',
        },
```
in backend with URL `http://localhost:7181/#app=HINPassword;test=test`.

**Note:** Backward-incompatible change (of none-documented feature).

### CODE REVIEW

**Comment by oetiker** [comment]
I would suggest to call the url data functionality from within the loadData function, and thus merge any data coming from the server with the data given on the url before calling setData ... in this way any formReconf will get called as if the data came from the backend, and no strange things will happen from calling setData twice.

**Comment by zaucker** [comment]
> I would suggest to call the url data functionality from within the loadData function, and thus merge any data coming from the server with the data given on the url before calling setData ... in this way any formReconf will get called as if the data came from the backend, and no strange things will happen from calling setData twice.
> 
> 

But then it will be called on every `_loadData()` and not just on appear. I guess this would probably work as the URL parameters are removed upon processing the first time.
Is this what you want?

**Comment by oetiker** [comment]
no, the reason I am saying it is that loadData may cause formReconf and this may depend on the data entered in the form ... and if it gets filled with server data first, the reaction might be undesirable ... 

hence I would merge the data

but yes guarding against repeated action is certainly a valid thing ... 

the reason the 'original' code worked was because I never used it in connection with existing data

**Comment by zaucker** [comment]
I left the callback in the appear handler, but merge the backend and urlData before setting the form.

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 1]
```diff
@@ -320,7 +332,7 @@ qx.Class.define("callbackery.ui.plugin.Form", {
                 that._loading--;
             },'getPluginData',this._cfg.name,'allFields',parentFormData,{ currentFormData: this._form.getData()});
         },
-        _loadData: function(){
+        _loadData: function(callback){
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:335]:
```suggestion
        _getUrlData: function () {
                    let data    = {};
                    let gotData = false;
                    let config  = callbackery.data.Config.getInstance();
                    this._urlFormElements.forEach(urlFormElement => {
                        let urlValue = config.getUrlConfigValue(urlFormElement.urlKey);
                        if (urlValue) {
                            data[urlFormElement.formKey] = urlValue;
                            // only do it once for the time being
                            config.removeUrlConfigEntry(urlFormElement.urlKey);
                            gotData = true;
                        }
                    });
                    return data;
                }
         }
        _loadData: function(mergeUrlData){
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 2]
```diff
@@ -334,6 +346,16 @@ qx.Class.define("callbackery.ui.plugin.Form", {
...
+                    let urlData;
+                    if (callback) {
+                        urlData = callback();
+                        if (!data) {
+                            data = urlData;
+                        }
+                        else {
+                            Object.assign(data, urlData);
+                        }
+                    }
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:358]:
```suggestion
                    if (getUrlData) {
                        urlData = this._getUrlData();
                        if (!data) {
                            data = urlData;
                        }
                        else {
                            Object.assign(data, urlData);
                        }
                    }
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 3]
```diff
@@ -36,22 +43,26 @@ qx.Class.define("callbackery.ui.plugin.Form", {
...
+                    this._urlFormElements.forEach(urlFormElement => {
+                        let urlValue = config.getUrlConfigValue(urlFormElement.urlKey);
+                        if (urlValue) {
+                            data[urlFormElement.formKey] = urlValue;
+                            // only do it once for the time being
+                            config.removeUrlConfigEntry(urlFormElement.urlKey);
+                            gotData = true;
+                        }
+                    });
+                    return data;
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:61]:
```suggestion
            this._loadData(true);
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 4]
```diff
@@ -36,22 +43,26 @@ qx.Class.define("callbackery.ui.plugin.Form", {
...
                 }
-            }
-            if (gotData) {
-                this._form.setData(data,true);
-                //this._reconfForm();
-            }
+            );
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:63]:
```suggestion
```

**Review by oetiker — CHANGES_REQUESTED** [review]
ich wuerde nach moeglichkeit versuchen etwas weniger inline code zu machen ...

**Comment by zaucker** [comment]
> ich wuerde nach moeglichkeit versuchen etwas weniger inline code zu machen ...

Suggestions comitted.

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 5]
```diff
@@ -320,7 +316,23 @@ qx.Class.define("callbackery.ui.plugin.Form", {
...
+                        if (urlValue) {
+                            data[urlFormElement.formKey] = urlValue;
+                            // only do it once for the time being
+                            config.removeUrlConfigEntry(urlFormElement.urlKey);
+                            gotData = true;
+                        }
+                    });
+                    return data;
+                }
+         }
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:334]:
```suggestion
         },
```

File: lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js [hunk 6]
```diff
@@ -36,22 +43,10 @@ qx.Class.define("callbackery.ui.plugin.Form", {
...
-            let data = {};
-            let gotData = false;
-            let urlCfg = callbackery.data.Config.getInstance().getUrlConfig();
-            for (let key in urlCfg){
-                let match = key.match(/^set_(.+)/);
-                if (match !== null){
-                    data[match[1]] = urlCfg[key];
-                    gotData = true;
+            this._loadData(true);
                 }
```
Comment (oetiker) [inline lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/plugin/Form.js:47]:
```suggestion
```

**Review by oetiker — APPROVED** [review]

**Pull request merged**